File: apps/api/src/main/java/com/lifttrack/api/domain/models/PersonalRecord.java
package com.lifttrack.api.domain.models;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

public record PersonalRecord(
        UUID uuid,
        UUID userUuid,
        UUID exerciseUuid,
        String recordType,
        BigDecimal weight,
        Integer reps,
        LocalDate achievedAt,
        UUID workoutSessionExerciseSetUuid,
        Instant createdAt) {

    public PersonalRecord create(UUID userUuid, UUID exerciseUuid, String recordType, BigDecimal weight, Integer reps,
            LocalDate achievedAt) {
        if (userUuid == null) {
            throw new IllegalArgumentException("User UUID cannot be null");
        }
        if (exerciseUuid == null) {
            throw new IllegalArgumentException("Exercise UUID cannot be null");
        }
        if (recordType == null || recordType.isBlank()) {
            throw new IllegalArgumentException("Record type cannot be null or blank");
        }
        if (achievedAt == null) {
            throw new IllegalArgumentException("Achieved at date cannot be null");
        }

        return new PersonalRecord(null, userUuid, exerciseUuid, recordType, weight, reps, achievedAt, null, null);
    }

}

File: apps/web/src/domain/models/personal-record.ts
export interface PersonalRecord {
  readonly uuid: string
  readonly userUuid: string
  readonly exerciseUuid: string
  readonly recordType: string
  readonly weight: number | null
  readonly reps: number | null
  readonly achievedAt: string
  readonly workoutSessionExerciseSetUuid: string | null
  readonly createdAt: string
}

export interface CreatePersonalRecord {
  readonly exerciseUuid: string
  readonly recordType: string
  readonly weight?: number
  readonly reps?: number
  readonly achievedAt: string
  readonly workoutSessionExerciseSetUuid?: string
}
